### PR TITLE
Privacy preserving image similarity

### PR DESCRIPTION
**Pull request by miriyalamani**
- Add privacy_protocol and user_email to embedding storage for cross-agency attribution
- Generate plain and anonymized embeddings for each image
- Add enable_anonymized toggle for privacy mode
- Anonymize with CLIPSeg ONNX (Xenova/clipseg-rd64-refined) before private embedding and PDQ hashing
- Remove torch; use onnxruntime and NumPy sigmoid
- Load CLIPSeg ONNX and processor files locally (no auto-download)
- Keep plain and private embeddings isolated in search and cache lookups
- Use Request pooler_output from SigLIP2 ONNX to improve combined CLIP+PDQ results
- Document dual ingestion, PDQ behavior, and PDQ-only single-series folder requirement

### CODE REVIEW

File: src/image-similarity/image_similarity/anonymizer.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    )
+    inputs = {k: v.to(clipseg.device) for k, v in inputs.items()}
+
+    with torch.inference_mode():
+        outputs = clipseg.model(**inputs)
+
+    logits = outputs.logits  # (num_labels, H, W)
+    for i, label in enumerate(labels):
+        mask_logits = logits[i]
+        mask_prob = torch.sigmoid(mask_logits).cpu().numpy()
```
Comment (nb950) [inline src/image-similarity/image_similarity/anonymizer.py:99]: do not use torch , its a huge module import, 
code should use onnxruntime exclusively and drop torch --this is  preferred for rescuebox plugins
hint :apply sigmoid using native NumPy instead of torch.sigmoid()

File: src/image-similarity/image_similarity/anonymizer.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+from PIL import Image, ImageFilter
+from transformers import CLIPSegProcessor, CLIPSegForImageSegmentation
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_TARGET_LABELS: list[str] = ["face", "person", "text", "sign", "logo"]
+DEFAULT_THRESHOLD = 0.3
+DEFAULT_DILATE = 15
+DEFAULT_BLUR = 5
+_CLIPSEG_MODEL_NAME = "CIDAS/clipseg-rd64-refined"
```
Comment (nb950) [inline src/image-similarity/image_similarity/anonymizer.py:30]: why is onnx model not  used here ?
[CIDAS/clipseg-rd64-refined ](https://huggingface.co/Xenova/clipseg-rd64-refined) does have onnx files

File: src/image-similarity/image_similarity/main.py [hunk 3]
```diff
@@ -628,6 +671,131 @@ def _build_metadata(
...
+    session: Session,
+    file_paths: list[str],
+    path_to_hash: dict[str, str],
+    ort_session: ort.InferenceSession,
+    processor: AutoImageProcessor,
+    user_email: str,
+    model_name: str = _DEFAULT_MODEL,
+) -> str:
+    """Anonymize images via CLIPSeg and store private embeddings + PDQ hashes."""
+    from image_similarity.anonymizer import anonymize_image, DEFAULT_TARGET_LABELS
```
Comment (nb950) [inline src/image-similarity/image_similarity/main.py:721]: move import to top of the file

File: src/image-similarity/image_similarity/main.py [hunk 4]
```diff
@@ -628,6 +671,131 @@ def _build_metadata(
...
+            f"First error: {failures[0][1]}"
+        )
+    return protocol
+
+
+def _backfill_private_pdq_hashes(
+    session: Session, file_paths: list[str], model_name: str, protocol: str
+) -> None:
+    """Fill in empty pdq_hash on existing private rows using the anonymized image."""
+    from image_similarity.anonymizer import anonymize_image
```
Comment (nb950) [inline src/image-similarity/image_similarity/main.py:772]: move import to top of the file

File: src/image-similarity/pyproject.toml [hunk 5]
```diff
@@ -17,6 +17,7 @@ sqlmodel = "*"
 sqlalchemy = "*"
 pgvector = "*"
 pdqhash = "*"
+torch = "*"
```
Comment (nb950) [inline src/image-similarity/pyproject.toml:20]: remove torch , please use onnxruntime where possible

File: src/image-similarity/README.md [hunk 6]
```diff
@@ -19,14 +19,18 @@ If you're looking for a concept like "people eating" rather than a specific scen
...
+
+**SigLIP2** (~1.7 GB) — vision encoder for embeddings:
 
 ```bash
 mkdir -p src/image-similarity/image_similarity/onnx_models
 curl -L -o src/image-similarity/image_similarity/onnx_models/siglip2-so400m-patch14-384.onnx \
   https://huggingface.co/onnx-community/siglip2-so400m-patch14-384-ONNX/resolve/main/onnx/vision_model.onnx
 ```
 
+**CLIPSeg** (~300 MB) — text-prompted segmentation for privacy-preserving anonymization. Auto-downloaded from HuggingFace on first use (cached in `~/.cache/huggingface/`).
```
Comment (nb950) [inline src/image-similarity/README.md:32]: do not use "Auto download" for model use onnx model file instead

**Review by nb950 — COMMENTED** [review]
please help fix the comments about using onnx model and remove torch

File: src/image-similarity/image_similarity/anonymizer.py (hunk 2 above)
Comment (miriyalamani) [reply src/image-similarity/image_similarity/anonymizer.py]: changed to onnx , used torch earlier because I was using sam3 and onnx files were too large

File: .gitignore [hunk 7]
```diff
@@ -28,6 +28,7 @@ MANIFEST
 *.bat
 *.exe
 *.onnx
+*.onnx.data
```
Comment (nb950) [inline .gitignore]: is onnx.data needed during runtime ?

File: poetry.lock [hunk 8]
```diff
@@ -1,4 +1,4 @@
-# This file is automatically @generated by Poetry 2.1.1 and should not be changed by hand.
+# This file is automatically @generated by Poetry 2.3.3 and should not be changed by hand.
```
Comment (nb950) [inline poetry.lock:1]: why did this file get updated when there is no change to toml file ?

i updated poetry with pipx install poetry==2.3.3
and now the lock file did update 
nothing to fix

**Comment by nb950** [comment]
please provide a google drive link of onnx model related files to download during runtime

File: src/image-similarity/image_similarity/app-info.md [hunk 9]
```diff
@@ -60,9 +73,10 @@ If you're looking for a concept like "people eating" rather than a specific scen
 ## How It Works (brief)
 
 1. Scan the input directory; for each image, check if its embedding already exists in `image_similarity_embeddings` (by path or content SHA-256). Only compute and store new vectors for files not already in the database.
-2. Compute PDQ perceptual hashes for all images (backfilling any that are missing).
-3. Look up or compute the **query image's** embedding and PDQ hash.
-4. Rank **only** the directory images using the selected scoring mode, return **top-k** results.
+2. **Dual ingestion:** For each image, create both a **plain** embedding and PDQ hash (`privacy_protocol = ""`) and a **private** embedding and PDQ hash (`privacy_protocol = "clipseg-blackout-v1"`). Private values use the same CLIPSeg mask: faces, text, logos, and other selected labels are blacked out before both encoding and PDQ hashing.
+3. Backfill missing PDQ hashes for existing embeddings.
+4. Look up or compute the **query image's** embedding and PDQ hash.
+5. Rank **only** the directory images using the selected scoring mode against plain embeddings, or private embeddings and masked-image PDQ hashes when anonymization is selected, then return **top-k** results.
```
Comment (nb950) [inline src/image-similarity/image_similarity/app-info.md]: its not clear to me on how this works , can you break down the steps further and or give an example ?

**Review by nb950 — APPROVED** [review]
